code/notificationbuilder/src/test/java/com/adobe/marketing/mobile/notificationbuilder/internal/builders/ManualCarouselNotificationBuilderTest.kt

```diff
@@ -0,0 +1,239 @@
+import org.robolectric.RuntimeEnvironment
+import org.robolectric.annotation.Config
+
+@RunWith(RobolectricTestRunner::class)
+@Config(sdk = [31])
+class ManualCarouselNotificationBuilderTest {
+
+    private lateinit var context: Context
+    private lateinit var pushTemplate: ManualCarouselPushTemplate
+    private lateinit var expandedLayout: RemoteViews
+    private lateinit var trackerActivityClass: Class<out Activity>
+    private lateinit var broadcastReceiverClass: Class<out BroadcastReceiver>
+
+    @Before
+    fun setUp() {
+        context = RuntimeEnvironment.getApplication()
+        pushTemplate = provideMockedManualCarousalTemplate(false)
+        trackerActivityClass = mockkClass(Activity::class, relaxed = true).javaClass
+        broadcastReceiverClass = mockkClass(BroadcastReceiver::class, relaxed = true).javaClass
+        mockkObject(PushTemplateImageUtils)
+        mockkObject(BasicNotificationBuilder)
+        mockkObject(PendingIntentUtils)
+        expandedLayout = mockkClass(RemoteViews::class)
+    }
+
+    @Test
+    fun `construct returns NotificationCompat Builder for valid inputs`() {
+        val result = ManualCarouselNotificationBuilder.construct(
+            context,
+            pushTemplate,
+            trackerActivityClass,
+            broadcastReceiverClass
+        )
+        assertNotNull(result)
+    }
+
+    @Test
+    fun `construct returns BasicNotificationBuilder if download image count is less than 3`() {
+        every { cacheImages(any()) } answers { 2 }
+        ManualCarouselNotificationBuilder.construct(
+            context,
+            pushTemplate,
+            trackerActivityClass,
+            broadcastReceiverClass
+        )
+        verify {
+            BasicNotificationBuilder.fallbackToBasicNotification(
+                context,
+                trackerActivityClass,
+                broadcastReceiverClass,
+                pushTemplate.data
+            )
+        }
+    }
+
+    @Test
+    fun `construct returns AEPPushNotificationBuilder if download image count is greater than equal to 3`() {
+        every { cacheImages(any()) } answers { 3 }
+        ManualCarouselNotificationBuilder.construct(
+            context,
+            pushTemplate,
+            trackerActivityClass,
+            broadcastReceiverClass
+        )
+        verify(exactly = 0) {
+            BasicNotificationBuilder.fallbackToBasicNotification(
+                context,
+                trackerActivityClass,
+                broadcastReceiverClass,
+                pushTemplate.data
+            )
+        }
+    }
+
+    @Test
+    fun `test downloadCarouselItems return non empty list for it retrieve the image bitmap from cache successfully`() {
+        every { getCachedImage(any()) } answers { mockkClass(Bitmap::class) }
+        val imagesList =
+            ManualCarouselNotificationBuilder.downloadCarouselItems(pushTemplate.carouselItems)
+        assertFalse(imagesList.isEmpty())
```

Comment:
can we verify each of the carousel items returned? additionally, carousel items are returned by `downloadCarouselItems` so the val on line 121 should be carouselItems or carouselItemsList to better reflect the function return value.

Reply:
That will better @rymorale 
I've updated the test, thanks for pointing out